src/ripple/consensus/Consensus.h

```diff
@@ -1102,12 +1133,119 @@ Consensus<Adaptor>::phaseOpen()
             openTime_.read(),
             idleInterval,
             adaptor_.parms(),
+            adaptor_.already(previousLedger_.seq()),
             j_))
     {
         closeLedger();
     }
 }
 
+template <class Adaptor>
+bool
+Consensus<Adaptor>::shouldPause() const
+{
+    auto const& parms = adaptor_.parms();
+    std::uint32_t const desynced (previousLedger_.seq() -
+        std::min(adaptor_.getValidLedgerIndex(), previousLedger_.seq()));
+    auto quorumKeys = adaptor_.getQuorumKeys();
+    auto const& quorum = quorumKeys.first;
+    auto& trustedKeys = quorumKeys.second;
+    std::size_t const totalValidators = trustedKeys.size();
+    std::size_t laggards = adaptor_.laggards(previousLedger_.seq(),
+        trustedKeys);
+
+    std::stringstream vars;
+    vars << " (working seq: " << previousLedger_.seq() << ", "
+         << "validated seq: " << adaptor_.getValidLedgerIndex() << ", "
+         << "am validator: " << adaptor_.validator() << ", "
+         << "have validated: " << adaptor_.haveValidated() << ", "
+         << "roundTime: " << result_->roundTime.read().count() << ", "
+         << "max consensus time: " << parms.ledgerMAX_CONSENSUS.count() << ", "
+         << "validators: " << totalValidators << ", "
+         << "laggards: " << laggards << ", "
+         << "offline: " << trustedKeys.size() << ", "
+         << "quorum: " << quorum << ", ";
+
+    if (!desynced ||
+        !laggards ||
+        !totalValidators ||
+        !adaptor_.validator() ||
+        !adaptor_.haveValidated() ||
+        result_->roundTime.read() > parms.ledgerMAX_CONSENSUS)
+    {
+        j_.debug() << "not pausing" << vars.str();
+        return false;
+    }
+
+    bool willPause = false;
+
+    /** Maximum phase with distinct thresholds to determine how
+     *  many validators must be on our same ledger sequence number.
+     *  The threshold for the 1st phase is >= the minimum number that
+     *  can achieve quorum. Threshold for the maximum phase is 100%
+     *  of all trusted validators. Progression from min to max phase is
+     *  simply linear. If there are 5 phases (maxPausePhase = 4)
+     *  and minimum quorum is 80%, then thresholds progress as follows:
+     *  1: >=80%
+     *  2: >=85%
+     *  3: >=90%
+     *  4: >=95%
+     *  5: =100%
+     *
+     *  With 3 phases (maxPausePhase = 2):
+     *  1: >=80%
+     *  2: >=90%
+     *  3: =100%
+     */
+    constexpr static std::size_t maxPausePhase = 2;
+
+    std::size_t const phase = desynced % (maxPausePhase + 1);
```

Comment:
Yes, I've commented accordingly.

Reply:
I don't have an example yet, but my gut is worried that cycling back around could increase dephasing between validators that are struggling to synchronize. I think its somewhat safe in that it might cause people to slow down even more on average, but I wonder if there are some edge cases where it actually increases time to recovery.

Reply:
Short of knowing why different % levels of peers on the correct sequence still might not be validating, I don't think there's really any way to assess this. Plus, something else is probably wrong if 100% of visible validators still are not enough to bring about consensus.

Reply:
Right, thats why I'm curious why you cycle versus stick at 100% (as @ximinez also asks below).

Reply:
@bachase said:
> One overall concern is the solution is specific, but fairly arbitrary, .e.g why 10s? why 4 phases? > My gut is worried about whether this overfits to a known example, but generalizes poorly.

10s should be safely less than 20s such that other pausing validators won't be interpreted as being offline: if their cycle aside from pausing exceeds the other 10s, then there is probably something wrong with them, so marking them offline is likely appropriate. I don't know if there's a way to optimize these values further but if you feel strongly that some other values are appropriate, let me know.

Reply:
Nothing extra I have in mind.